File: tests/conformance/evidence.mjs
// Conformance evidence for JavaScript tests. A test that runs a shared fixture case records the
// feature it proves, the fixture, the runtime and the case. Nothing is written unless
// CRUDUI_CONFORMANCE_EVIDENCE names a directory; scripts/check-conformance.mjs reads it.
import { appendFileSync, mkdirSync } from 'node:fs';
import path from 'node:path';

/**
 * Record one case result.
 *
 * @param {{feature: string, fixture: string, runtime: string, case: string, passed: boolean}} item
 */
export function recordConformance({ feature, fixture, runtime, case: name, passed }) {
  for (const [key, value] of Object.entries({ feature, fixture, runtime, case: name })) {
    if (typeof value !== 'string' || value === '') throw new TypeError(`Conformance evidence ${key} must be a non-empty string`);
  }
  if (typeof passed !== 'boolean') throw new TypeError('Conformance evidence passed must be a boolean');
  const directory = process.env.CRUDUI_CONFORMANCE_EVIDENCE;
  if (!directory) return;
  mkdirSync(directory, { recursive: true });
  const line = JSON.stringify({ feature, fixture, runtime, case: name, passed });
  appendFileSync(path.join(directory, `javascript-${process.pid}.jsonl`), `${line}\n`);
}

/**
 * Run one fixture case and record its result for every feature it proves; the case's own
 * failure is rethrown.
 */
export async function provesConformance({ features, fixture, runtime, case: name }, run) {
  let passed = false;
  try {
    const result = await run();
    passed = true;
    return result;
  } finally {
    for (const feature of features) recordConformance({ feature, fixture, runtime, case: name, passed });
  }
}

File: packages/validator-go/validator/internal/conformance/evidence.go
// Package conformance records conformance evidence for Go tests.
//
// A test that runs a shared fixture case records the feature it proves, the
// fixture, the runtime and the case. Nothing is written unless the environment
// variable CRUDUI_CONFORMANCE_EVIDENCE names a directory;
// scripts/check-conformance.mjs reads it. The format matches
// tests/conformance/evidence.mjs.
package conformance

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"testing"
)

// runtime is the runtime key of this validator in contracts/features.json.
const runtime = "go"

// evidence is one recorded line; field order is the line's key order.
type evidence struct {
	Feature string `json:"feature"`
	Fixture string `json:"fixture"`
	Runtime string `json:"runtime"`
	Case    string `json:"case"`
	Passed  bool   `json:"passed"`
}

// Record registers a cleanup on t that records whether the fixture case name
// passed. Call it at the start of the subtest that runs exactly that case, so
// the result reflects the case and every subtest it runs. Empty arguments fail
// the test.
func Record(t testing.TB, feature, fixture, name string) {
	t.Helper()
	for key, value := range map[string]string{"feature": feature, "fixture": fixture, "case": name} {
		if value == "" {
			t.Fatalf("conformance evidence %s must be a non-empty string", key)
		}
	}
	t.Cleanup(func() {
		directory := os.Getenv("CRUDUI_CONFORMANCE_EVIDENCE")
		if directory == "" {
			return
		}
		if err := write(directory, evidence{feature, fixture, runtime, name, !t.Failed()}); err != nil {
			t.Errorf("conformance evidence: %v", err)
		}
	})
}

// write appends one JSON line to <directory>/go-<pid>.jsonl.
func write(directory string, item evidence) error {
	line, err := json.Marshal(item)
	if err != nil {
		return err
	}
	if err := os.MkdirAll(directory, 0o755); err != nil {
		return err
	}
	path := filepath.Join(directory, fmt.Sprintf("%s-%d.jsonl", runtime, os.Getpid()))
	file, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return err
	}
	if _, err := file.Write(append(line, '\n')); err != nil {
		file.Close()
		return err
	}
	return file.Close()
}
